docs/website/contents/for-developers/utxo-hd/utxo-hd.md

```diff
@@ -0,0 +1,436 @@
+values that are used when dealing with later blocks.
+
+
+### Managing the differences
+
+To ensure the properties of the Ouroboros protocols, the Consensus layer needs
+to be able to perform rollbacks on the chain of at most `k` blocks (which in
+mainnet equals `2160` blocks). Because of this, the differences of the last `k`
+blocks cannot be considered immutable and therefore they cannot yet be flushed to
+persistent storage. This same principle is the one that dictates that ledger
+snapshots (for restarting the node) have to store ledger states before or at the
+immutable tip of the chain.
+
+Following this same reasoning, the way differences are carried around changes
+depending on the specific backend used by the LedgerDB, whether it lives on the
+disk or in memory:
+
+#### On-disk backend
+
+The on-disk backend uses the concept of an _anchor_ which is before or at the
+immutable tip. This _anchor_ contains a full UTXO set stored in the disk, in what we call the `BackingStore`. In
+order to get values for applying a particular block, the Consensus layer has to
+read those values from the anchored UTXO set and apply all the differences from
+that point to the tip of the chain.
+
+This means that to the pre-UTXO-HD LedgerDB that held the last `k` ledger
+states, a side sequence is added which holds the differences resulted from
+applying each of the last `k` blocks. This sequence is held in a `FingerTree`
+which contains the combination of all the differences that can be applied
+directly to a set of values.
+
+The Consensus layer will periodically flush the differences in between the
+anchor and the current immutable tip to the on-disk backend, advancing the
+chain.
+
+#### In-memory backend
+
+The in-memory backend augments each of the `k` values contained in the LedgerDB
+to hold a full UTXO set. This emulates exactly how the LedgerDB looked like
+before UTXO-HD. After each operation with the Ledger, the resulting differences
+are applied to the set of values on the tip, producing the new UTXO set.
+
+The memory footprint of this solution is almost equivalent to the pre-UTXO-HD
+one. There aren't `k` UTXO sets, but just by Haskell's sharing, there is one
+UTXO set, the others sharing most of their contents with each one's predecessor.
+
+### The forker abstraction
+
+In order to perform operations with the Ledger layer, Consensus defines the
+`Forker` abstraction as _an API to evaluate forks_. Forkers give access to
+reading values at a specific point in the chain. Its implementation depends on
+which backend is at use in the LedgerDB abstracting over both of them.
+
+It is important to note that when using the on-disk backend, a `Forker` will
+mantain a consistent view of the _anchored_ UTXO set, which means that writes to
+the anchor are queued while the `Forker` is held. For this reason, `Forker`s
+should be an ephemeral resource, released as soon as possible.
+
+### Interacting with the Ledger layer (examples)
```

Comment:
I'm not convinced that we should include these diagrams. My thinking is that if we want diagrams, then we should separate them into diagrams for the internals of the LedgerDB, and a single diagram for block application up to the LedgerDB API. We could leave out the diagram for the internals of the LedgerDB, since the API is abstract and so one could argue that we don't need to know the internals to understand UTXO-HD